nexus_constructor/component_fields.py

```diff
@@ -0,0 +1,190 @@
+        self.field_type_combo.setMinimumWidth(100)
+        fix_horizontal_size = QSizePolicy()
+        fix_horizontal_size.setHorizontalPolicy(QSizePolicy.Fixed)
+        self.field_type_combo.setSizePolicy(fix_horizontal_size)
+
+        self.value_type_combo = QComboBox()
+        self.value_type_combo.addItems(list(DATASET_TYPE.keys()))
+
+        self.value_line_edit = QLineEdit()
+        self.value_line_edit.setValidator(
+            FieldValueValidator(self.field_type_combo, self.value_type_combo)
+        )
+        self.value_line_edit.validator().is_valid.connect(
+            partial(
+                validate_line_edit,
+                self.value_line_edit,
+                tooltip_on_accept="Value is cast-able to numpy type.",
+                tooltip_on_reject="Value is not cast-able to selected numpy type.",
+            )
+        )
+        self.value_line_edit.validator().validate(self.value_line_edit.text(), 0)
+
+        self.nx_class_combo = QComboBox()
+
+        self.edit_button = QPushButton("Edit")
+        edit_button_size = 50
+        self.edit_button.setMinimumWidth(edit_button_size)
+        self.edit_button.setMaximumWidth(edit_button_size)
+        self.edit_button.setSizePolicy(fix_horizontal_size)
+        self.edit_button.clicked.connect(self.show_edit_dialog)
+
+        self.layout = QHBoxLayout()
+        self.layout.addWidget(self.field_name_edit)
+        self.layout.addWidget(self.field_type_combo)
+        self.layout.addWidget(self.value_line_edit)
+        self.layout.addWidget(self.nx_class_combo)
+        self.layout.addWidget(self.edit_button)
+        self.layout.addWidget(self.value_type_combo)
+
+        self.layout.setAlignment(Qt.AlignLeft)
+        self.setLayout(self.layout)
+
+        self.setFrameShadow(QFrame.Raised)
+        self.setFrameShape(QFrame.StyledPanel)
+
+        # Allow selecting this field widget in a list by clicking on it's contents
+        self.field_name_edit.installEventFilter(self)
+        self.value_line_edit.installEventFilter(self)
+        self.edit_button.installEventFilter(self)
+        self.nx_class_combo.installEventFilter(self)
+        self.value_type_combo.highlighted.connect(self.something_clicked)
+        self.field_type_combo.highlighted.connect(self.something_clicked)
+
+        # Set the layout for the default field type
+        self.field_type_changed()
+
+    @property
+    def field_type(self):
+        return FieldType(self.field_type_combo.currentText())
+
+    @property
+    def name(self):
+        return self.field_name_edit.text()
+
+    @property
+    def dtype(self):
+        if self.field_type == FieldType.scalar_dataset:
+            return self.value.dtype
+        if self.field_type == FieldType.array_dataset:
+            return np.array(self.value).dtype
+
+    @property
+    def value(self):
+        if self.field_type == FieldType.scalar_dataset:
```

Comment:
Having all these equality checks in this class is messy and it's only going to grow. I wonder if you could have a class that describes the part of the field that changes depending on it's type and have a inheritance structure that describes the different behaviour for each type? Or do you think that'll be messier?

Reply:
I think that's a good idea, I'm just not sure the best way to go about doing it. Wouldn't you still be doing equality checks on the type of the class regardless of whether you use inheritance? I like the idea of separating them out as i think the field widget has too much responsibility currently.

Reply:
I agree that might be the way to go when there are more `FieldType`s to deal with, but I think it will be easier to get the implementation of that right when that is the case rather than trying to do it now. I suggest the current implementaiton is less complex at this stage.

Reply:
Sure, sounds sensible.